profiling/src/internal/observation/observations.rs

```diff
@@ -7,43 +7,58 @@ use super::super::Sample;
 use super::trimmed_observation::{ObservationLength, TrimmedObservation};
 use crate::internal::Timestamp;
 use std::collections::HashMap;
+use crate::collections::identifiable::*;
+use lz4_flex::frame::FrameEncoder;
+use std::io::Write;
 
 struct NonEmptyObservations {
     aggregated_data: HashMap<Sample, TrimmedObservation>,
-    timestamped_data: Vec<TrimmedTimestampedObservation>,
+    compressed_timestamped_data: FrameEncoder<Vec<u8>>,
```

Comment:
How much do we save by compressing vs just using an array of values, and indicies into there?

Reply:
We don't actually need to index into the array (which is why compressing it works).

On the "how much do we save", here's some back-of-the-napkin numbers:

* Size of each observation: 4 bytes stacktrace, 4 bytes labels, 8 bytes timestamp, 8 bytes * N profile types
* 100 threads * 100 samples per second * 60 seconds * (4 + 4 + 8 + 8 * 4 profile types enabled for Ruby by default) = 28 800 000 bytes

* For... reasons... the test app ends up recording data for 103 threads so I get
  * Uncompressed 29 MiB
  * Compressed 4 MiB

Which seems like a nice improvement as well. This data is highly compressible, (lots of small numbers, zeros, numbers next to each other, ...) so we could make a smarter uncompressed representation, but I think the compressor takes very well care of that for us.

Reply:
Have you tried compressing in a Struct of Arrays format, and compress each array of fields of the `TrimmedTimestampedObservation` struct independently?
I'd think that this would give you a better compression ratio since the data would be more homogenous

Reply:
Also, you might get a better compression ratio using the [Linked](https://docs.rs/lz4_flex/latest/lz4_flex/frame/enum.BlockMode.html#variant.Linked) block mode since you're doing a lot of small write and thus the block size will default to the minimum (64KB)

Reply:
I think those are very valid options; having said that, I'm not convinced they are what we would want here:

> Have you tried compressing in a Struct of Arrays format, and compress each array of fields of the TrimmedTimestampedObservation struct independently?
I'd think that this would give you a better compression ratio since the data would be more homogenous

If I'm understanding your suggestion, doing this would mean having multiple `FrameEncoder`s, with multiple underlying buffers:

```
struct NonEmptyObservations {
    stack_ids: FrameEncoder<Vec<u8>>,
    label_ids: FrameEncoder<Vec<u8>>,
    timestamps: FrameEncoder<Vec<u8>>,
    values: FrameEncoder<Vec<u8>>,
    // ...
```

This would mean more allocations, especially when growing the backing vecs -- there's no longer one single vec that gets doubled, but multiple small ones. :thinking:

I suspect (without having tried it, to be fair xD) this would cause more heap fragmentation.

> Also, you might get a better compression ratio using the [Linked](https://docs.rs/lz4_flex/latest/lz4_flex/frame/enum.BlockMode.html#variant.Linked) block mode since you're doing a lot of small write and thus the block size will default to the minimum (64KB)

I'm assuming that if the encoder is referencing previous blocks, it means that it's doing more work (rather than e.g. looking only at the current block).

Depending on the win, If it's a few % I'm not sure it's worth doing more work on the profiling sample write path vs the memory savings, since anyway it's not a lot of data.